tools/logger/logger.c

```diff
@@ -171,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
 	return 0;
 }
 
+/* Thanks to inotify, wait 5 minutes for 'expected_file' to appear in
+ * 'watched_dir'.  Then return fopen(expected_files). fopen() failures are NOT
+ * handled; e.g. NULL from fopen() is just passed through.
+ */
+static FILE *wait_fopen(const char *watched_dir, const char *expected_file)
+{
+	if (access(watched_dir, R_OK)) {
+		fprintf(stderr, "error: %s() cannot read %s\n", __func__, watched_dir);
+		return NULL;
+	}
+
+	const int iqueue = inotify_init();
+	const int dwatch = inotify_add_watch(iqueue, watched_dir, IN_CREATE);
+	FILE *fstream;
+
+	char * const fpath = malloc(strlen(watched_dir) + 1 + strlen(expected_file) + 1);
```

Comment:
can we have the declarations first?

Reply:
Why, do we need to support any legacy, pre-C99 compiler here? Leaving variables uninitialized makes the code more verbose and less safe: not possible to use `const` and more risk of errors. 
This is not the Linux kernel here, no need to support a baroque bad habit that never even existed in any other programming language.

https://stackoverflow.com/a/4105334/317623

Reply:
It is just that Linux coding standard is the native language for so many of us here. The Linux coding standard does not forbid stack variable initializations, just mixing declarations (with or without initialization) and code, so it is possible satisfy both the security and the dinosaurs with more code blocks. Just saying, this is fine by me as it is.

Reply:
The Linux standard was stuck to C89 which forbid C99 mixing. It was recently upgraded to C11.

Leaving both security and dinosaurs aside for a minute, how would you rewrite this without mixing and without losing a few `const`? `{ }` blocks can often help but here these constants and variables are used in the entire function. So I'm afraid some constants would have to be downgraded to variables, no?

Reply:
I do not see the problem, am I missing something?

`static void open_file(const char *watched_dir, const char *expected_file)
{
	const int iqueue = inotify_init();
	const int dwatch = inotify_add_watch(iqueue, watched_dir, IN_CREATE);
	struct stat expected_stat;
	void *ret_stream = NULL;
	char * const fpath = malloc(strlen(watched_dir) + 1 + strlen(expected_file) + 1);

...
	return ret_stream;
}


static void *wait_open(const char *watched_dir, const char *expected_file)
{
	if (access(watched_dir, R_OK)) {
		fprintf(stderr, "error: %s() cannot read %s\n", __func__, watched_dir);
		return NULL;
	}

	return open_files(watched_dir, expected_file);
}`

Reply:
Agreed, you can always add extra functions until mixing is avoided, artificially creating as many blocks as you need. However having a tiny "public" function that merely checks whether an argument exists and then delegates all the rest to another function that actually does all the work is an awkward solution for a C89 problem that does not exist anymore.

BTW many people don't even understand this old C89 rule, in the past I had review comments asking me not to do this:
```C
f(...)
{
   statement1;
    { int i;

    }
}

```

Don't get me wrong: giant functions are bad too but that's off topic.